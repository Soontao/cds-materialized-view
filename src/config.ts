import { cwdRequireCDS } from "cds-internal-tool";


const cds = cwdRequireCDS();

export const materializedConfig = {

  /**
   * mtxs t0 name
   */
  get t0(): string {
    return cds.env.get("requires.multitenancy.t0") ?? "t0";
  },

  /**
   * the interval between check tenant is update/onboard/off-board
   */
  get tenantCheckInterval() {
    return (cds.env.get("materialized.check.tenant.interval") ?? 60) * 1000;
  },

  /**
   * the interval between check each view is refresh or not
   */
  get viewCheckInterval() {
    return (cds.env.get("materialized.check.view.interval") ?? 1) * 1000;
  },

  /**
   * the default interval between each view to trigger refresh behavior
   */
  get defaultViewRefreshInterval() {
    return (cds.env.get("materialized.default.view.refresh.interval") ?? 3600) * 1000;
  },

};
