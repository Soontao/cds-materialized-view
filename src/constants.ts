
export const ANNOTATIONS = {
  CDS_MATERIALIZED_VIEW: "@cds.materialized.view",
  CDS_MATERIALIZED_INTERVAL: "@cds.materialized.interval",
};
