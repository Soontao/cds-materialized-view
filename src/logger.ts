import { cwdRequireCDS } from "cds-internal-tool";

export function getLogger() {
  const cds = cwdRequireCDS();
  return cds.log("materialize");
}