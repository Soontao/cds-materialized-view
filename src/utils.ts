import { cwdRequireCDS } from "cds-internal-tool";


export function deepClone<T = any>(obj: T): T {
  if (obj === undefined) {
    return undefined as T;
  }
  if (obj === null) {
    return null as T;
  }
  return JSON.parse(JSON.stringify(obj));
}

export const MATERIALIZED_VIEW_PREFIX = "materialized.";

export function privilegedUser() {
  const cds = cwdRequireCDS();
  return new cds.User.Privileged("materialized-view-job");
}
