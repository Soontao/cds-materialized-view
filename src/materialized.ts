import { Definition } from "cds-internal-tool";
import { materializedConfig } from "./config";
import { ANNOTATIONS } from "./constants";
import { MATERIALIZED_VIEW_PREFIX } from "./utils";

/**
 * check the entity definition is materialized view or not
 * 
 * @param def 
 * @returns 
 */
export function isMaterializedView(def: Definition): boolean {
  if (def.query === undefined && def.projection === undefined) {
    return false;
  }
  if (def[ANNOTATIONS.CDS_MATERIALIZED_VIEW] !== true) {
    return false;
  }

  return true;
}

/**
 * format given view name to materialized view name
 * 
 * @param name 
 * @returns 
 */
export function getMaterializedViewName(name: string) {
  if (name.startsWith(MATERIALIZED_VIEW_PREFIX)) {
    return name;
  }
  return MATERIALIZED_VIEW_PREFIX + name;
}

export function getMaterializedViewRefreshInterval(def: Definition): number {
  return (def[ANNOTATIONS.CDS_MATERIALIZED_INTERVAL] ?? materializedConfig.defaultViewRefreshInterval) * 1000;
}
