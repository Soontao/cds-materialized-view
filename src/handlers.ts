/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable camelcase */
/* eslint-disable max-len */
/* eslint-disable @typescript-eslint/ban-ts-comment */
import {
  CSN, cwdRequireCDS, DatabaseService, EntityDefinition,
  isCDSRequest, NextFunction, Request, VarDefinition
} from "cds-internal-tool";
import { materializedConfig } from "./config";
import { HTTP_HEADER_X_REFRESH_AT, TABLE_MATERIALIZED_REFRESH_JOB } from "./constants";
import { getLogger } from "./logger";
import { getMaterializedViewName, isMaterializedView } from "./materialized";
import { deepClone, privilegedUser } from "./utils";


async function csn4(tenant?: string): Promise<CSN> {
  const cds = cwdRequireCDS();
  const { "cds.xt.ModelProviderService": mp } = cds.services;
  return mp!.getCsn({ tenant, toggles: ["*"], activated: true }); // REVISIT: ['*'] should be the default
}

export async function refreshMaterializedInfo(data: any, req: Request<{ tenant: string, options: any }>) {
  const logger = getLogger();
  const { tenant, options } = req.data;
  if (tenant === materializedConfig.t0) {
    return;
  }
  const cds = cwdRequireCDS();

  // 'deploy' for new subscription while 'upgrade'|'extend' for existed tenant
  const csn = req.event === "deploy" ? (options?.csn ?? await csn4()) : await csn4(tenant);

  const localViews = Object.entries<EntityDefinition>(csn.definitions)
    .filter(([, def]) => isMaterializedView(def))
    .map(([name]) => ({
      view: name,
      refreshAt: "1900-01-01T00:00:00.000Z",
      nextRefreshAt: "1900-01-01T00:00:00.000Z",
    }));

  // after transaction
  req.on("succeeded", async () => {
    const { INSERT, DELETE, SELECT } = cds.ql;
    await cds.tx({ tenant, user: privilegedUser() }, async (tx) => {
      const dbViews: Array<{ view: string }> = await tx.run(SELECT.from(TABLE_MATERIALIZED_REFRESH_JOB).columns("view"));

      /**
       * view existed in db but not in CSN
       */
      const cleanViews = dbViews.filter(dbView => localViews.find(view => view.view === dbView.view) === undefined);
      if (cleanViews.length > 0) {
        logger.info(cleanViews.length, "deprecated materialized view detected, clean them into meta");
        await tx.run(DELETE.from(TABLE_MATERIALIZED_REFRESH_JOB).where({ view: { in: cleanViews } }));
      }

      /**
       * view existed in CSN but not in db
       */
      const newViews = localViews.filter(view => dbViews.find(dbView => dbView.view === view.view) === undefined);
      if (newViews.length > 0) {
        logger.info(localViews.length, "fresh materialized view detected, writing them into meta");
        await tx.run(INSERT.into(TABLE_MATERIALIZED_REFRESH_JOB).entries(...newViews));
      }
    });
  });

}


// TODO: maybe custom builder instead of re-write query

/**
 * rewrite Query for materialized view
 *  
 * @param req 
 * @returns 
 */
export function rewriteQueryForMaterializedView(req: Request) {
  const cds = cwdRequireCDS();
  const { query } = req;

  // @ts-ignore
  if (
    cds.context instanceof cds.Request &&
    cds.context?._?.req?.headers?.["x-cds-materialized-view-disable"] === "true"
  ) {
    return;
  }
  // TODO: maybe req.query could be a string
  // @ts-ignore
  if (typeof query !== "object" || query.SELECT.from.ref.length !== 1 || typeof query?.SELECT?.from?.ref?.[0] !== "string") {
    return;
  }
  // @ts-ignore
  if (!(query.SELECT.from.ref[0] in cds.db.model.definitions)) {
    return;
  }
  // @ts-ignore
  if (!isMaterializedView(cds.db.model.definitions[query.SELECT.from.ref[0]])) {
    // TODO: should use tenant model
    return;
  }
  // @ts-ignore
  const entityName = req.query.SELECT.from.ref[0];
  const viewName = getMaterializedViewName(entityName);

  // @ts-ignore
  req.query.SELECT.from.ref[0] = viewName;

  (req as any).__materialized__ = { entityName, viewName };
}

/**
 * append 
 * 
 * @param this 
 * @param req 
 * @param next 
 * @returns 
 */
export async function appendRefreshAtHeader(this: DatabaseService, req: Request, next: NextFunction) {
  const cds = cwdRequireCDS();
  const { SELECT } = cds.ql;
  if (isCDSRequest(cds.context) && materializedConfig.addRefreshAtHeader) {
    if (typeof (req as any).__materialized__ === "object") {
      const { entityName } = (req as any).__materialized__;
      const r = await this.run(SELECT.one.from(TABLE_MATERIALIZED_REFRESH_JOB).where({ view: entityName }));
      if (r !== null && typeof r.refreshAt === "string") {
        cds.context._.res.header(HTTP_HEADER_X_REFRESH_AT, r.refreshAt);
      }
    }

  }
  return next();
}

export const metaTenantEntities = {
  materialized_refresh_job: {
    kind: "entity",
    elements: {
      view: { key: true, type: "cds.String", length: 255 },
      nextRefreshAt: { type: "cds.Timestamp" },
      refreshAt: { type: "cds.Timestamp" },
    }
  }
};

/**
 * rewrite/enhance tenant CSN with materialized view (table) entities
 * 
 * @param csn 
 * @returns 
 */
export function rewriteAfterCSNRead(csn: CSN) {
  const cds = cwdRequireCDS();
  const logger = getLogger();

  let hasMaterializedView = false;
  for (const [name, def] of Object.entries<VarDefinition>(csn.definitions)) {

    if (!isMaterializedView(def)) {
      continue;
    }

    if (def.kind !== "entity") {
      logger.warn("definition", name, "is not a view, skip processing");
      continue;
    }

    if (def.query === undefined && def.projection === undefined) {
      logger.warn("definition", name, "is a raw entity, skip processing");
      continue;
    }

    const newDef = {
      name: getMaterializedViewName(name),
      kind: "entity",
      elements: deepClone(def.elements)
    } as EntityDefinition;

    for (const [name, ele] of Object.entries(newDef.elements)) {
      if (ele.type === undefined) {
        throw cds.error(`there is not type definition for element ${newDef.name}.${name}`);
      }
    }

    csn.definitions[newDef.name!] = newDef;
    hasMaterializedView = true;
    logger.debug("append materialized view", newDef);

  }

  if (hasMaterializedView) {
    csn.definitions = Object.assign(csn.definitions, metaTenantEntities);
  }

  return csn;
}
