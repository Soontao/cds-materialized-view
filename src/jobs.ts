/* eslint-disable camelcase */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/ban-ts-comment */
import { concurrency } from "@newdash/newdash/concurrency";
import { CSN, cwdRequireCDS, EntityDefinition, Service } from "cds-internal-tool";
import { materializedConfig } from "./config";
import { TABLE_MATERIALIZED_REFRESH_JOB } from "./constants";
import { getLogger } from "./logger";
import { getMaterializedViewName, getMaterializedViewRefreshInterval } from "./materialized";
import { MetaRefreshJob } from "./types";
import { privilegedUser } from "./utils";

export async function refreshMaterializedViews() {

  const cds = cwdRequireCDS();
  const logger = getLogger();
  const { SELECT } = cds.ql;
  const { "cds.xt.ModelProviderService": mps } = cds.services;

  try {

    // TODO: cache tenant id
    const tenants: Array<{ ID: string }> = await cds.tx(
      { tenant: materializedConfig.t0, user: privilegedUser() },
      tx => tx.run(SELECT.from("cds.xt.Tenants").columns("ID")),
    );

    const tenantIds = tenants.map(({ ID }) => ID);

    const results = await Promise.allSettled(tenantIds.map(async tenant => {

      // TODO: cache csn maybe
      const csn: CSN = await mps!.tx(
        {
          tenant: materializedConfig.t0,
          user: privilegedUser()
        },
        (tx: any) => tx.getCsn({ tenant, toggles: ["*"], activated: true })
      );

      await cds.tx(
        { tenant, user: privilegedUser() },
        async (tx) => {
          // FOR UPDATE LOCK
          // REVISIT: no wait maybe
          const views: Array<MetaRefreshJob> = await tx.run(
            SELECT
              .from(TABLE_MATERIALIZED_REFRESH_JOB)
              .where({ nextRefreshAt: { "<=": new Date().toISOString() } })
              .forUpdate()
          );

          if (views.length > 0) {
            const results = await Promise.allSettled(views.map(view => refreshSingleView(view, csn, tx)));

            for (const result of results) {
              if (result.status === "rejected") {
                logger.error("refresh view in", tenant, "failed", result.reason);
              }
            }
          }
        }
      );

    }));

    for (const result of results) {
      if (result.status === "rejected") {
        logger.error("refresh materialized view failed", result.reason);
      }
    }
  }
  catch (error) {
    logger.error("un-expected error when refresh materialized views", error);
  }


}
const refreshSingleView = concurrency.limit(
  async function refreshSingleView(view: MetaRefreshJob, csn: any, tx: Service) {
    const cds = cwdRequireCDS();
    const { DELETE, INSERT, UPDATE, SELECT } = cds.ql;

    const viewName = view.view;
    const materializedViewName = getMaterializedViewName(viewName);
    const def: EntityDefinition = csn.definitions[view.view];
    const interval = getMaterializedViewRefreshInterval(def);

    await tx.run(DELETE.from(materializedViewName));
    // @ts-ignore
    await tx.run(INSERT.into(materializedViewName).as(SELECT.from(viewName)));


    // update next refresh time
    await tx.run(
      UPDATE
        .entity(TABLE_MATERIALIZED_REFRESH_JOB)
        .where({ view: view.view })
        .set({
          refreshAt: new Date().toISOString(),
          nextRefreshAt: new Date(Date.now() + interval).toISOString()
        })
    );

  },
  materializedConfig.viewRefreshConcurrency,
);

/**
 * jobs handles
 */
export const jobs = new Set<NodeJS.Timer>();

/**
 * setup jobs
 */
export function setupJobs() {

  const logger = getLogger();
  if (materializedConfig.jobEnabled) {

    // JOB: refresh view content
    jobs.add(
      setInterval(
        refreshMaterializedViews,
        materializedConfig.viewCheckInterval * 1000
      )
    );

  }
  else {
    logger.info("materialized view refresh job is not enabled");
  }

}

/**
 * stop all background jobs
 */
export function clearJobs() {
  for (const job of jobs) {
    clearInterval(job);
  }
}
