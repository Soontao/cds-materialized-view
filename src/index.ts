/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable max-len */
/* eslint-disable camelcase */
/* eslint-disable @typescript-eslint/ban-ts-comment */
import { cwdRequireCDS } from "cds-internal-tool";
import { clearJobs, setupJobs } from "./jobs";
import { getLogger } from "./logger";
import { rewriteAftreCSNRead, rewriteQueryForMaterializedView } from "./rewrite";

const cds = cwdRequireCDS();

cds.once("served", () => {

  const { "cds.xt.DeploymentService": ds, "cds.xt.ModelProviderService": mps } = cds.services;

  if (ds === undefined) {
    getLogger().error("cds.xt.DeploymentService is not enabled, cds-materialized-view feature is disabled");
    return;
  }

  // REWRITE: tenant CSN for tenant onboard/upgrade
  mps.prepend((mps: any) => { mps.after("getCsn", rewriteAftreCSNRead); });

  // REWRITE: database query
  cds.db.prepend(db => { db.before("READ", rewriteQueryForMaterializedView); });

  setupJobs();

  // TODO: ensure only have one instance to refresh materialized view

});


export { clearJobs };
