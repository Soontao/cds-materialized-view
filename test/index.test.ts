import { cwdRequireCDS, setupTest } from "cds-internal-tool";
import { HTTP_HEADER_X_REFRESH_AT } from "../src/constants";
import { sleep } from "./utils";

describe("Main Test Suite", () => {

  const axios = setupTest(__dirname, "./app")

  const cds = cwdRequireCDS()

  cds.env.materialized = {
    check: {
      tenant: { interval: 1 },
      view: { interval: 1 },
    }
  }

  axios.defaults.auth = { username: 'alice', password: '' }

  const { clearJobs } = require("../src/jobs")

  it('should subscribe new tenant', async () => {
    // subscribe tenant 1
    await axios.put('/-/cds/saas-provisioning/tenant/t1', {
      "subscribedTenantId": "t1",
      "subscribedSubdomain": "subdomain1",
      "eventType": "CREATE"
    })
    await sleep(2000)
  });

  it('should support get metadata', async () => {
    const { data } = await axios.get("/app/$metadata")
    expect(data).toMatch(/UniqPeopleNames/)
  });

  it('should support query objects', async () => {
    const { data } = await axios.get("/app/UniqPeopleNames")
    expect(data).toMatchSnapshot()
  });

  it('should support query objects with refresh at header', async () => {
    const { headers: h1 } = await axios.get("/app/UniqPeopleNames")
    expect(h1[HTTP_HEADER_X_REFRESH_AT]).toBeUndefined()

    const { headers: h2 } = await axios.get("/app/UniqPeopleNames", {
      headers: {
        [HTTP_HEADER_X_REFRESH_AT]: 'true'
      }
    })
    expect(h2[HTTP_HEADER_X_REFRESH_AT]).toBeDefined()
    expect(isNaN(new Date(h2[HTTP_HEADER_X_REFRESH_AT] as string).getTime())).toBeFalsy()
  });

  it('should support filter', async () => {
    const { data } = await axios.get("/app/UniqPeopleNames?$filter=Name ne 'Theo 2'")
    expect(data).toMatchSnapshot()
  });

  it('should support aggregation', async () => {
    const { data } = await axios.get("/app/MaxHousePrice")
    expect(data).toMatchSnapshot()
  });

  it('should support un-subscribe', async () => {
    await axios.delete('/-/cds/saas-provisioning/tenant/t1')
  });

  afterAll(async () => {
    clearJobs()
    await sleep(1000)
  })

});
