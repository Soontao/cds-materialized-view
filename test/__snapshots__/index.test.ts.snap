// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Test Suite should support aggregation for projection 1`] = `
{
  "@odata.context": "$metadata#AvgHousePrice",
  "value": [
    {
      "avgPrice": 671400,
    },
  ],
}
`;

exports[`Main Test Suite should support aggregation for select 1`] = `
{
  "@odata.context": "$metadata#MaxHousePrice",
  "value": [
    {
      "maxPrice": 2357853,
    },
  ],
}
`;

exports[`Main Test Suite should support filter 1`] = `
{
  "@odata.context": "$metadata#UniqPeopleNames",
  "value": [
    {
      "Name": "Theo First",
    },
    {
      "Name": "Theo Second",
    },
  ],
}
`;

exports[`Main Test Suite should support non mapping for projection 1`] = `
{
  "@odata.context": "$metadata#FullNameHouse",
  "value": [
    {
      "FullAddress": "c1,p1,Chengdu,s1",
      "ID": "97dc99ee-f61e-11ec-b939-0242ac120002",
      "price": 100000,
    },
    {
      "FullAddress": "c1,p2,Beijing,s3",
      "ID": "97dc9bc4-f61e-11ec-b939-0242ac120002",
      "price": 204324,
    },
    {
      "FullAddress": "c1,p3,Shanghai,s4",
      "ID": "97dc9cb4-f61e-11ec-b939-0242ac120002",
      "price": 2357853,
    },
    {
      "FullAddress": "c1,p4,Guilin,s5",
      "ID": "97dc9da4-f61e-11ec-b939-0242ac120002",
      "price": 23423,
    },
  ],
}
`;

exports[`Main Test Suite should support query objects 1`] = `
{
  "@odata.context": "$metadata#UniqPeopleNames",
  "value": [
    {
      "Name": "Theo 2",
    },
    {
      "Name": "Theo First",
    },
    {
      "Name": "Theo Second",
    },
  ],
}
`;
